Duplicati/Library/Backend/Telegram/TelegramBackend.cs

```diff
@@ -0,0 +1,475 @@
+            {
+                throw new UserInformationException(Strings.NoApiHashError, nameof(Strings.NoApiHashError));
+            }
+
+            if (string.IsNullOrEmpty(m_phoneNumber))
+            {
+                throw new UserInformationException(Strings.NoPhoneNumberError, nameof(Strings.NoPhoneNumberError));
+            }
+
+            if (string.IsNullOrEmpty(m_channelName))
+            {
+                throw new UserInformationException(Strings.NoChannelNameError, nameof(Strings.NoChannelNameError));
+            }
+
+            InitializeTelegramClient(m_apiId, m_apiHash, m_phoneNumber);
+        }
+
+        private static void InitializeTelegramClient(int apiId, string apiHash, string phoneNumber)
+        {
+            var tmpTelegramClient = m_telegramClient;
+            m_telegramClient = new TelegramClient(apiId, apiHash, m_sessionStore, phoneNumber);
+            tmpTelegramClient?.Dispose();
+        }
+
+        public void Dispose()
+        {
+            // Do not dispose m_telegramClient because of "session-based" authentication.
+        }
+
+        public string DisplayName { get; } = Strings.DisplayName;
+        public string ProtocolKey { get; } = "https";
+
+        public IEnumerable<IFileEntry> List()
+        {
+            return SafeExecute<IEnumerable<IFileEntry>>(() =>
+                {
+                    Authenticate();
+                    EnsureChannelCreated();
+                    var fileInfos = ListChannelFileInfos();
+                    var result = fileInfos.Select(fi => fi.ToFileEntry());
+                    return result;
+                },
+                nameof(List));
+        }
+
+        public Task PutAsync(string remotename, string filename, CancellationToken cancelToken)
+        {
+            return SafeExecute(() =>
+                {
+                    var channel = GetChannel();
+                    using (var fs = File.OpenRead(filename))
+                    {
+                        var fsReader = new StreamReader(fs);
+
+                        cancelToken.ThrowIfCancellationRequested();
+                        EnsureConnected();
+                        var file = m_telegramClient.UploadFile(remotename, fsReader, cancelToken).GetAwaiter().GetResult();
+
+                        cancelToken.ThrowIfCancellationRequested();
+                        var inputPeerChannel = new TLInputPeerChannel {ChannelId = channel.Id, AccessHash = (long)channel.AccessHash};
+                        var fileNameAttribute = new TLDocumentAttributeFilename
+                        {
+                            FileName = remotename
+                        };
+
+                        EnsureConnected();
+                        m_telegramClient.SendUploadedDocument(inputPeerChannel, file, remotename, "application/zip", new TLVector<TLAbsDocumentAttribute> {fileNameAttribute}, cancelToken).GetAwaiter().GetResult();
+                        fs.Close();
```

Comment:
I don't think it's necessary to explicitly call `Close`.  The call to `Dispose` should take care of it.

Reply:
Fixed, thanks